proposals/proposals

```diff
@@ -0,0 +1,214 @@
+> - The set of candidate methods is reduced to contain only methods from the most derived types: For each method `C.F` in the set, where `C` is the type in which the method `F` is declared, all methods declared in a base type of `C` are removed from the set. Furthermore, if `C` is a class type other than `object`, all methods declared in an interface type are removed from the set.  
+>   > *Note*: This latter rule only has an effect when the method group was the result of a member lookup on a type parameter having an effective base class other than `object` and a non-empty effective interface set. *end note*
+> - If the resulting set of candidate methods is empty, then further processing along the following steps are abandoned, and instead an attempt is made to process the invocation as an extension method invocation ([§12.8.9.3](expressions.md#12893-extension-method-invocations)). If this fails, then no applicable methods exist, and a binding-time error occurs.
+> - **The resulting set of candidate methods is grouped by ***overload_resolution_priority***. All candidates not in the highest (by standard integer comparison) ***overload_resolution_priority*** group are removed from the set.**
+> - The best method of the set of candidate methods is identified using the overload resolution rules of [§12.6.4](expressions.md#1264-overload-resolution). If a single best method cannot be identified, the method invocation is ambiguous, and a binding-time error occurs. When performing overload resolution, the parameters of a generic method are considered after substituting the type arguments (supplied or inferred) for the corresponding method type parameters.
+
+As an example, this feature would cause the following code snippet to print "Span", rather than "Array":
+
+```cs
+using System.Runtime.CompilerServices;
+
+var d = new C1();
+int[] arr = [1, 2, 3];
+d.M(arr); // Prints "Span"
+
+class C1
+{
+    [OverloadResolutionPriority(1)]
+    public void M(ReadOnlySpan<int> s) => Console.WriteLine("Span");
+    // Default overload resolution priority
+    public void M(int[] a) => Console.WriteLine("Array");
+}
+```
+
+The effect of this change is that, like pruning for most-derived types, we add a final pruning for overload resolution priority. Because this pruning occurs at the very end of the overload resolution
+process, it does mean that a base type cannot make its members higher-priority than any derived type. This is intentional, and prevents an arms-race from occuring where a base type may try to always
+be better than a derived type. For example:
+
+```cs
+using System.Runtime.CompilerServices;
+
+var d = new Derived();
+d.M([1, 2, 3]); // Prints "Derived", because members from Base are not considered due to finding an applicable member in Derived
```

Comment:
I understand why this is done, but it is a shame that there isn't a way to allow this to occur.

For example, we could support the following:
```csharp
using System.Runtime.CompilerServices;

var d = new Derived();
d.M([1, 2, 3]); // Prints "Base", because member in Derived is treated as if it were defined in Base, therefore the ROS overload wins

class Base
{
    [OverloadResolutionPriority(1)]
    public void M(ReadOnlySpan<int> s) => Console.WriteLine("Base");
}

class Derived : Base
{
    [OverloadResolutionPriority(0, TreatAsBase = typeof(Base))]
    public void M(int[] a) => Console.WriteLine("Derived");
}
```

I would find this option useful (I've run into this issue a few times), but I understand that it might be considered unnecessarily complicated.